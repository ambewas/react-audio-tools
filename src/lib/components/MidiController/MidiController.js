import { Component } from "react";
import PropTypes from "prop-types";
import { midiMessages } from "../../constants";
import webmidi from "webmidi";


class MidiController extends Component {
  state = {}
  static propTypes = {
    children: PropTypes.func,
    controlled: PropTypes.bool,
  }

  constructor(props) {
    super(props);

    if (props.controlled) {
      this.requestMidiAccess();
    }
  }

  requestMidiAccess = () => {
    webmidi.enable(err => {
      if (err) {
        alert("No midi support in your browser."); // eslint-disable-line
      } else {
        this.initializeMidi();
      }
    });
  }

  initializeMidi = () => {
    const inputs = webmidi.inputs;

    for (let input in inputs) {
      inputs[input].addListener("noteon", "all", this.handleNoteOn);
      inputs[input].addListener("noteoff", "all", this.handleNoteOff);
    }
  }

  handleNoteOn = (e) => {
    this.setState({
      midiMsg: {
        type: midiMessages.NOTE_ON,
        pitch: e.note.number,
        velocity: e.rawVelocity,
      },
    });
  }

  handleNoteOff = (e) => {
    this.setState({
      midiMsg: {
        type: midiMessages.NOTE_OFF,
        pitch: e.note.number,
        velocity: 0,
      },
    });
  }

  render() {
    const { children } = this.props;
    const { midiMsg } = this.state;

    return (
      children(midiMsg)
    );
  }
}

export default MidiController;
