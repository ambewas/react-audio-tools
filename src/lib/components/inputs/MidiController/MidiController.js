import { Component } from "react";
import PropTypes from "prop-types";
import { midiMessages } from "../../../constants";
import webmidi from "webmidi";


class MidiController extends Component {
  state = {}
  static propTypes = {
    children: PropTypes.func,
  }

  constructor(props) {
    super(props);
  }

  componentDidMount() {
    this.requestMidiAccess();
  }

  componentWillUnmount() {
    if (this.canAccessMidi) {
      for (let input in this.inputs) {
        this.inputs[input].removeListener("noteon", "all", this.handleNoteOn);
        this.inputs[input].removeListener("noteoff", "all", this.handleNoteOff);
      }
      webmidi.disable();
    }
  }

  requestMidiAccess = () => {
    webmidi.enable(err => {
      if (err) {
        this.canAccessMidi = false;
        alert("No midi support in your browser."); // eslint-disable-line
      } else {
        this.canAccessMidi = true;
        this.initializeMidi();
      }
    });
  }

  initializeMidi = () => {
    this.inputs = webmidi.inputs;
    for (let input in this.inputs) {
      this.inputs[input].addListener("noteon", "all", this.handleNoteOn);
      this.inputs[input].addListener("noteoff", "all", this.handleNoteOff);
    }
  }

  handleNoteOn = (e) => {
    this.setState({
      midiMsg: {
        type: midiMessages.NOTE_ON,
        pitch: e.note.number,
        velocity: e.rawVelocity,
      },
    });
  }

  handleNoteOff = (e) => {
    this.setState({
      midiMsg: {
        type: midiMessages.NOTE_OFF,
        pitch: e.note.number,
        velocity: 0,
      },
    });
  }

  render() {
    const { children } = this.props;
    const { midiMsg } = this.state;

    return (
      children(midiMsg)
    );
  }
}

export default MidiController;
