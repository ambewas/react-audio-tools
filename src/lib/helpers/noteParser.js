export const noteParser = pitch => {
  if (pitch < 0 || pitch > 127) {
    return undefined;
  }

  const octave = Math.floor(pitch / 12);
  const notes = ["C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B"];
  const noteIndex = pitch % 12;

  return notes[noteIndex] + octave;
};
