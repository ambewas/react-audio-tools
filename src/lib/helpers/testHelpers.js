import React from "react";
import ReactTestRenderer from "react-test-renderer";
import { shallow } from "enzyme";

export const setupComponent = (Component, defaultProps) => propOverrides =>{
  const props = {
    ...defaultProps,
    ...propOverrides,
  };

  const wrapper = ReactTestRenderer.create(
    <Component {...props} />
  ).toJSON();

  const component = shallow(<Component {...props} />);

  return {
    props,
    wrapper,
    component,
  };
};

