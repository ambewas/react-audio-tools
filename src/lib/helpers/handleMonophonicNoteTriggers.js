import { midiMessages } from "../constants";
import { append, without, last } from "ramda";
import { noteParser } from "./noteParser";

// handle midi messages for the specific audioNode class that is using the function.
// note that we're assigning statics on the class, to remember which notes have been
// played before. Got to store them somewhere. On the class itself seems fine for now.
export const handleMonophonicNoteTriggers = (midiMsg, someClass) => {
  const { type, pitch, velocity } = midiMsg;

  const parsedPitch = noteParser(pitch);

  if (type === midiMessages.SUSTAIN) {
    if (velocity === 127) {
      someClass.sustained = true;
    } else {
      someClass.sustained = false;
    }
  }

  // save which notes are currently playing based on the midi msg
  if (type === midiMessages.NOTE_ON) {
    someClass.playingNotes = append(parsedPitch, someClass.playingNotes);
  }

  if (type === midiMessages.NOTE_OFF) {
    someClass.playingNotes = without(parsedPitch, someClass.playingNotes);
  }

  const someNoteIsDepressed = someClass.playingNotes && someClass.playingNotes.length > 0;

  // trigger the audioNode for the last note in playingNotes (it's monophonic after all);
  if (someNoteIsDepressed) {
    const noteToTrigger = last(someClass.playingNotes);

    if (noteToTrigger !== someClass.lastPlayingNote) {
      someClass.audioNode.triggerAttack(noteToTrigger);
    }
    someClass.lastPlayingNote = noteToTrigger;
  }

  // if nothing is depressed anymore & sustain is off, kill the audioNode
  if (!someNoteIsDepressed && !someClass.sustained) {
    someClass.audioNode.triggerRelease();
    someClass.lastPlayingNote = undefined;
  }
};
