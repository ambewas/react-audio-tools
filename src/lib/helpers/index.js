import { handleMonophonicNoteTriggers } from "./handleMonophonicNoteTriggers";
import { noteParser } from "./noteParser";
import { setupComponent } from "./testHelpers";

export {
  handleMonophonicNoteTriggers,
  noteParser,
  setupComponent,
};
