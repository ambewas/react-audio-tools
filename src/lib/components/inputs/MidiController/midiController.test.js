import { setupComponent } from "../../../helpers/index";
import MidiController from "./MidiController";
import webmidi from "webmidi";

const setup = setupComponent(MidiController, {
  children: () => null,
});

describe("MidiController", () => {
  it("should render without crashing", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  describe("without midi access", () => {
    beforeAll(() => {
      webmidi.enable = jest.fn(err => err({ someError: "someError" }));
    });

    it("should set a canAccessMidi static on the class to false if midi is not available", () => {
      const { component } = setup();
      const componentInstance = component.instance();

      expect(webmidi.enable).toBeCalled();
      expect(componentInstance.canAccessMidi).toBe(false);
    });
  });

  describe("with midi access", () => {
    beforeAll(() => {
      webmidi.enable = jest.fn(err => err());
    });

    it("should set a canAccessMidi static on the class to true if midi available", () => {
      const { component } = setup();
      const componentInstance = component.instance();

      expect(webmidi.enable).toBeCalled();
      expect(componentInstance.canAccessMidi).toBe(true);
    });
  });
});
