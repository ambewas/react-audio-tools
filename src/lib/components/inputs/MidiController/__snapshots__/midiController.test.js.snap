// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MidiController should render without crashing 1`] = `null`;
